fix(pose): nested transition fns

amcdnl opened the pull request:
Fixes #646

Andarist commented on packages/pose-core/src/factories/setter.ts:
```diff
@@ -97,6 +97,19 @@ const resolveTransition = <V, A>(
+    /**
+     * transition: () => { d: () => tweenFn }
+     */
+    if (typeof resolvedTransition === 'object') {
```
according to types I think it cant be non-object, so checking it here is a little bit pointless

Andarist replied:
also when we are talking about types - this change has to be reflected in the typings

Andarist commented:
Could you add a test for this too?

InventingWithMonster commented:
Yep happy to merge with some tests

amcdnl commented:
@Andarist - I removed the check. 2 questions -

1) It looks like the types do work since the `TransitionFactory` is loosely typed to return `any`.

```ts
export type Props = { [key: string]: any };

export type TransitionFactory<A> = (
  props: Props
) => TransitionDefinition | A | false;
```

2) I tried to write a test but its not working correctly. Maybe you can lend me a pointer. 

I added a new function to `testPoser` called `nestedFunctionalDefTransition`. It looks like:

```
  nestedFunctionalDefTransition: {
    x: 5,
    transition: () => ({
      x: () => 10
    })
  },
```
then in the `'resolves custom transitions correctly'` test I added a case after the `functionalDefTransition` that returned the same result but called my new pose instead like:

```
.then(() => {
      expect(testPoser.get().x).toBe(8 * 2)
      return testPoser.set('nestedFunctionalDefTransition')
    })
```

it fails the same with and without my fix to:

```
    expect(received).toBe(expected) // Object.is equality

    Expected: 16
    Received: 50

      244 |     })
      245 |     .then(() => {
    > 246 |       expect(testPoser.get().x).toBe(8 * 2)
          |                                 ^
      247 |       return testPoser.set('nestedFunctionalDefTransition')
      248 |     })
      249 |     .then(() => {

      at _tests/index.test.ts:246:33
```

Any ideas what I'm doing wrong?

Andarist commented:
1. Im not a TS expert, so im not entirely sure about what im going to say - but maybe we should change the types here, seems like `TransitionDefinition` is too loose, but probably we shouldnt touch it right now. It seems to me that this type is used too often in our definitions - i think it was supposed to be a type of a single transition property but its also used as return type of `TransitionFactory` which is semantically incorrect. I would change this and instead of this:
```ts
  transition?: TransitionMap<A> | TransitionFactory<A>;
``` 
do this 
```ts
type TransitionMapFactory<A> = (props: Props) => TransitionMap<A>;
// ...
  transition?: TransitionMap<A> | TransitionMapFactory<A>;
```
This would still not be ideal, but at least would be slightly more correct and it would make it easier to narrow those types in the future.

2. not sure, could u push out or changes? I could take a look at it.

amcdnl commented:
@Andarist - Changes are pushed.

@InventingWithMonster - Any thoughts on the types?

Andarist commented:
@amcdnl could u also push your failing tests? At the moment this PR changes only the implementation file.

Andarist commented:
This PR can be closed as those commits got merged in as part of the https://github.com/Popmotion/popmotion/pull/714
